File: agent/memory/failure_memory_matcher.py
from __future__ import annotations

from typing import Any


class FailureMemoryMatcher:
    @staticmethod
    def match(query: str, memories: list[dict[str, Any]]) -> dict[str, Any] | None:
        q_tokens = set(FailureMemoryMatcher._tokenize(query))
        if not q_tokens:
            return None

        best = None
        best_score = 0.0
        for item in memories:
            text = str(item.get("user_query", ""))
            if query in text or text in query:
                return {"memory": item, "similarity": 0.9}
            m_tokens = set(FailureMemoryMatcher._tokenize(text))
            if not m_tokens:
                continue
            inter = len(q_tokens.intersection(m_tokens))
            union = len(q_tokens.union(m_tokens))
            score = inter / union if union else 0.0
            if score > best_score:
                best_score = score
                best = item

        if best is None or best_score < 0.35:
            return None
        return {"memory": best, "similarity": round(best_score, 4)}

    @staticmethod
    def _tokenize(text: str) -> list[str]:
        text = text.lower().strip()
        if not text:
            return []
        tokens = [tok for tok in text.replace("，", " ").replace(",", " ").split() if tok]
        if not tokens:
            return [text]
        return tokens

File: agent-ts/src/memory/failure/core/FailureMemoryMatcher.ts
import type { FailureMemoryItem } from "../model/FailureMemoryItem.js";

export class FailureMemoryMatcher {
  match(query: string, memories: readonly FailureMemoryItem[]): { item: FailureMemoryItem; similarity: number } | undefined {
    const queryTokens = this.tokenize(query);
    if (queryTokens.size === 0) return undefined;
    let best: FailureMemoryItem | undefined;
    let bestScore = 0;
    for (const item of memories) {
      if (!item.userQuery) continue;
      if (query.includes(item.userQuery) || item.userQuery.includes(query)) {
        return { item, similarity: 0.9 };
      }
      const itemTokens = this.tokenize(item.userQuery);
      const intersection = [...queryTokens].filter((token) => itemTokens.has(token)).length;
      const union = new Set([...queryTokens, ...itemTokens]).size;
      const score = union === 0 ? 0 : intersection / union;
      if (score > bestScore) {
        best = item;
        bestScore = score;
      }
    }
    return best && bestScore >= 0.35 ? { item: best, similarity: Number(bestScore.toFixed(4)) } : undefined;
  }

  private tokenize(text: string): Set<string> {
    return new Set(text.toLowerCase().trim().split(/[\s,，。！？；：、]+/u).filter(Boolean));
  }
}
